proposals/csharp-7.0/tuples.md

```diff
@@ -0,0 +1,437 @@
+    ;
+    
+tuple_type_element
+    : type identifier?
+    ;
+```
+
+## Additions to [Types](../../spec/types.md)
+
+> The following sections should be added after [Nullable types](../../spec/types.md#nullable-types) (at the end of the current *Value types* section.)
+
+### Tuple types
+
+Tuple types are declared with the following syntax:
+
+```csharp
+public (int sum, int count) Tally(IEnumerable<int> values) { ... }
+
+var t = Tally(myValues);
+Console.WriteLine($"Sum: {t.sum}, count: {t.count}");
+ ```
+
+The syntax `(int sum, int count)` indicates an anonymous data structure with public fields of the given names and types, referred to as *tuple*.
+
+Tuple values can be created using *tuple literals*:
+
+```csharp
+var t1 = (sum: 0, count: 1);
+var t2 = (0, 1);     // field names are optional
+```
+
+Tuples cannot be created with the `new` operator. The `new` operator can be used to initialize arrays of tuples, or nullable tuples:
+
+```csharp
+var array = new (int x, int y)[10];
+var nullable = new (int x, int y)?();
+```
+
+Tuples can be created for a known target type:
+
+```csharp
+public (int sum, int count) Tally(IEnumerable<int> values)
+{
+    var s = 0; var c = 0;
+    foreach (var value in values) { s += value; c++; }
+    return (s, c); // target typed to (int sum, int count)
+}
+```
+
+Specifying field names is optional. Duplicate names are disallowed.
+
+```csharp
+var t1 = (sum: 0, count: 1); // OK, all fields are named
+var t2 = (sum: 0, 1);       // Ok, implies (int sum, int);
+var t3 = (sum: 0, sum: 1);  // error! duplicate names.
+```
+
+### Tuple literals
+
+```csharp
+var t1 = (0, 2);              // infer tuple type from values
+var t2 = (sum: 0, count: 1);  // infer tuple type from names and values
+```
+
+A tuple literal is "target typed" whenever possible. The tuple literal has a "conversion from expression" to any tuple type, as long as the element expressions of the tuple literal have an implicit conversion to the element types of the tuple type.
+
+```csharp
+(string name, byte age) t = (null, 5); // Ok: the expressions null and 5 convert to string and byte
+```
+
+In cases where the tuple literal is not part of a conversion, it acquires its "natural type", which means a tuple type where the element types are the types of the constituent expressions. Since not all expressions have types, not all tuple literals have a natural type either:
+
+```csharp
+var t = ("John", 5); // Ok: the type of t is (string, int)
+var t = (null, 5); //   Error: null doesn't have a type
+```
+
+A tuple literal may include names, in which case they become part of the natural type:
+
+```csharp
+var t = (name: "John", age: 5); // The type of t is (string name, int age)
+```
+
+Tuple literals may be deconstructed directly:
+
+```csharp
+(string x, byte y, var z) = (null, 1, 2);
+(string x, byte y) t = (null, 1);
+```
+
+Or for deconstructing assignment:
+
+```csharp
+string x;
+byte y;
+
+(x, y) = (null, 1);
+(x, y) = (y, x); // swap!
```

Comment:
This appears to be an assignment in which the left-hand-side isn't an lvalue, which would be an error.  How is it permitted?